src/D2L.CodeStyle.Analyzers/Diagnostics.cs

```diff
@@ -514,5 +514,77 @@ public static class Diagnostics {
+			category: "Immutability",
+			defaultSeverity: DiagnosticSeverity.Error,
+			isEnabledByDefault: true
+		);
+
+		public static readonly DiagnosticDescriptor ArraysAreMutable = new DiagnosticDescriptor(
+			id: "D2L0062",
+			title: "Arrays are mutable and thus can't be held by an immutable type.",
+			messageFormat: "{0}[]'s are mutable even if {0}'s are immutable because array items can be substituted. Prefer an immutable container type such as ImmutableArray<{0}>.",
+			category: "Immutability",
+			defaultSeverity: DiagnosticSeverity.Error,
+			isEnabledByDefault: true
+		);
+
+		public static readonly DiagnosticDescriptor DynamicObjectsAreMutable = new DiagnosticDescriptor(
+			id: "D2L0063",
+			title: "Dynamic objects are mutable and thus can't be held by an immutable type.",
+			messageFormat: "Dynamic objects are mutable and thus can't be held by an immutable type.",
+			category: "Immutability",
+			defaultSeverity: DiagnosticSeverity.Error,
+			isEnabledByDefault: true
+		);
+
+		public static readonly DiagnosticDescriptor TypeParameterIsNotKnownToBeImmutable = new DiagnosticDescriptor(
+			id: "D2L0064",
+			title: "Type parameter is not known to be immutable.",
+			messageFormat: "The type parameter {0} is not known to be immutable. Add [Immutable] to it.",
+			category: "Immutability",
+			defaultSeverity: DiagnosticSeverity.Error,
+			isEnabledByDefault: true
+		);
+
+		public static readonly DiagnosticDescriptor UnexpectedTypeKind = new DiagnosticDescriptor(
+			id: "D2L0065",
+			title: "Unsupported type kind inside an immutable type declaration.",
+			messageFormat: "The type kind {0} is unexpected/unhandled by the immutability analyzer, and for safety we are emitting an error. This is a bug in the analyzer.",
+			category: "Immutability",
+			defaultSeverity: DiagnosticSeverity.Error,
+			isEnabledByDefault: true
+		);
+
+		public static readonly DiagnosticDescriptor TypeNeedsImmutableAttribute = new DiagnosticDescriptor(
+			id: "D2L0066",
+			title: "Type may be mutable.",
+			messageFormat: "The {0} {1} is missing the [Immutable]{2} attribute, so it isn't safe to be held by an immutable type",
+			category: "Immutability",
+			defaultSeverity: DiagnosticSeverity.Error,
+			isEnabledByDefault: true
+		);
+
+		public static readonly DiagnosticDescriptor MemberIsNotReadOnly = new DiagnosticDescriptor(
+			id: "D2L0067",
+			title: "Member must be readonly inside immutable type.",
+			messageFormat: "The {0} {1} must be readonly because {2} is immutable.",
+			category: "Immutability",
+			defaultSeverity: DiagnosticSeverity.Error,
+			isEnabledByDefault: true
+		);
+
+		public static readonly DiagnosticDescriptor UnexpectedMemberKind = new DiagnosticDescriptor(
+			id: "D2L0068",
+			title: "Unsupported member type in immutable type declaration.",
+			messageFormat: "The member {0} is of type {1} which is not supported by the immutability analyzer. This is a bug.",
```

Comment:
This "shouldn't happen" (until it does).

Reply:
Consider telling users what they can do if they experience this bug (e.g. file an issue in this repo). [`DiagnosticDescriptor.HelpLinkUri`](https://docs.microsoft.com/en-us/dotnet/api/microsoft.codeanalysis.diagnosticdescriptor.helplinkuri) may be useful here, even if its just a link to the repo itself.

Reply:
Cool -- I deeplinked to the create issue page with a reasonable title for these two diagnostics.